_drafts/algebra-of-types.md

```diff
@@ -0,0 +1,319 @@
+  }
+}
+
+proveTypesAreIsomorphic[String + (Int + Boolean), (String + Int) + Boolean].check
+// + OK, passed 100 tests.
+```
+
+### Unit `a + 0 = a`
+
+For product the unit was the terminal object and for its dual the unit is the terminal object's dual: the initial object.
+The reasoning is the same, just the arrows are reversed.
+
+![a + ø as sum][a + ø as sum]
+
+Where `ø` denotes the initial object. We want to prove that a is a better candidate for a coproduct than `a + ø` is.
+
+![a as coproduct][a as coproduct]
+
+Again, we can only use the universal construction, for which we have to find our injections and a morphism from `a` to `a + ø`. And they are, again, right there!
+
+![a is a better coproduct candidate][a is a better coproduct candidate]
+
+Now regarding the code, we will show that `A + 0` is isomorphic to `A`. We declared `Nothing` to be our initial object, but, unfortunately, we would get stuck with a compile error[^2] if we used `Nothing` for our proof. 
+That's not a problem, we can come up with an initial object in no time, we just need an uninhabited type. An abstract final class will be perfect.
+
+```scala
+import org.scalacheck.{Arbitrary, Gen}
+
+abstract final class Zero
+implicit val arbitraryZero: Arbitrary[Zero] = Arbitrary(Gen.fail)
+
+implicit def unitOfSum[A] = new Isomorphism[A + Zero, A] {
+  def a2b = {
+    case Left(a) => a
+    case _ => throw new IllegalStateException("The world has just blown up, there is an instance of Zero")
+  }
+  def b2a = injectLeft
+}
+
+proveTypesAreIsomorphic[String + Zero, String].check
+// + OK, passed 100 tests.
+```
+
+## The distributive property `a * (b + c) = a * b + a * c`
+
+Now that one is interesting. My first reaction to this was "Noooo waay... but how? ...wow, that's cool"
+
+```scala
+implicit def distributivity[A, B, C] = new Isomorphism[A * (B + C), (A * B) + (A * C)] {
+  def a2b = {
+    case (a, Left(b)) => injectLeft((a, b))
+    case (a, Right(c)) => injectRight((a, c))
+  }
+
+  def b2a = {
+    case Left((a, b)) => (a, injectLeft(b))
+    case Right((a, c)) => (a, injectRight(c))
+  }
+}
+
+proveTypesAreIsomorphic[String * (Boolean + Int), (String * Boolean) + (String * Int)].check
+// + OK, passed 100 tests.
+```
+
+## Algebraic data types
+
+Algebraic data types are built from products and sums. The product do not have to be implemented as a pair, and the sum do not have to be an `Either`. We can come up with a product type like this: `case class Product[A, B](a: A, b: B)` and it is just as fine as the tuple we used before. For sum types, although [dotty has a Haskell-like syntax], in scala we use sealed traits for now. For example:
+
+```scala
+sealed trait Coproduct[A, B]
+case class One[A, B](a: A) extends Coproduct[A, B]
+case class Other[A, B](b: B) extends Coproduct[A, B]
+```
+
+Now we know why they are called algebraic.
+
+## Some extras
+
+Remember what was the first code snippet that was used to try the isomorphism in the [previous post]?
```

Comment:
Shouldn't this link work in the rendered version?

Reply:
I'm not sure, to be honest I have no idea why the images work, well, at least some of them :)

Reply:
thank you for your comments